File: src/main/java/nu/peg/slack/pt/di/binder/SlackBinder.java
package nu.peg.slack.pt.di.binder;

import nu.peg.slack.pt.api.slack.SlackApi;
import nu.peg.slack.pt.api.transport.TransportApi;
import nu.peg.slack.pt.di.factory.*;
import nu.peg.slack.pt.service.*;

import org.glassfish.hk2.utilities.binding.AbstractBinder;

public class SlackBinder extends AbstractBinder {

    @Override
    protected void configure() {
        boolean devMode = tryParseBoolean(System.getProperties().getOrDefault("developmentMode", null));

        if (devMode) {
            bindFactory(DevelopmentSlackApiFactory.class).to(SlackApi.class);
        } else {
            bindFactory(SlackApiFactory.class).to(SlackApi.class);
        }
        bindFactory(TransportApiFactory.class).to(TransportApi.class);
        bindFactory(OauthServiceFactory.class).to(OauthService.class);
        bindFactory(ConnectionServiceFactory.class).to(ConnectionService.class);
        bindFactory(LocationServiceFactory.class).to(LocationService.class);
        bindFactory(InteractiveServiceFactory.class).to(InteractiveService.class);
    }

    private boolean tryParseBoolean(Object obj) {
        if (null == obj || !(obj instanceof String)) return false;

        String booleanString = ((String) obj).toLowerCase();
        // returns false if it fails to parse
        return Boolean.parseBoolean(booleanString);
    }
}

File: src/main/kotlin/nu/peg/slack/pt/di/binder/SlackBinder.kt
package nu.peg.slack.pt.di.binder

import nu.peg.slack.pt.api.slack.SlackApi
import nu.peg.slack.pt.api.transport.TransportApi
import nu.peg.slack.pt.di.factory.*
import nu.peg.slack.pt.service.*

import org.glassfish.hk2.utilities.binding.AbstractBinder

class SlackBinder : AbstractBinder() {

    protected fun configure() {
        val devMode = tryParseBoolean((System.getProperties() as java.util.Map).getOrDefault("developmentMode", null))

        if (devMode) {
            bindFactory(DevelopmentSlackApiFactory::class.java).to(SlackApi::class.java)
        } else {
            bindFactory(SlackApiFactory::class.java).to(SlackApi::class.java)
        }
        bindFactory(TransportApiFactory::class.java).to(TransportApi::class.java)
        bindFactory(OauthServiceFactory::class.java).to(OauthService::class.java)
        bindFactory(ConnectionServiceFactory::class.java).to(ConnectionService::class.java)
        bindFactory(LocationServiceFactory::class.java).to(LocationService::class.java)
        bindFactory(InteractiveServiceFactory::class.java).to(InteractiveService::class.java)
    }

    private fun tryParseBoolean(obj: Any?): Boolean {
        if (null == obj || obj !is String) return false

        val booleanString = obj.toLowerCase()
        // returns false if it fails to parse
        return java.lang.Boolean.parseBoolean(booleanString)
    }
}
